# Data Cleaning for Customer Satisfaction Analysis

This notebook is part of a larger project exploring customer satisfaction in Brazilian e-commerce using the [Olist dataset](https://www.kaggle.com/olistbr/brazilian-ecommerce).  
In this first notebook, we focus on loading, exploring, and cleaning each raw data table individually to prepare them for further integration and analysis.

**Goals of this notebook:**
- Load and inspect the raw datasets
- Explore key features and structure of each table
- Handle missing values and duplicates
- Save clean, analysis-ready data for joining

**This notebook is followed by:**
- [Data Integration Notebook](02_data_integration_preprocessing.ipynb): joins cleaned tables into one dataset and performs preprocessing
- [Exploratory Analysis Notebook](./03_customer-satisfaction-analysis.ipynb): explores patterns in customer reviews and potential drivers of satisfaction

## **Structure of the Notebook**

*Note: Section links and “Back to top” links work best in Jupyter environments (e.g., Jupyter Lab or VS Code). They may not work as expected when clicked directly on GitHub.*

- [Data Overview](#data-overview)  
  - [Loading raw datasets](#loading-raw-datasets)
  - [Exploring datasets (structure, sampling, descriptive statistics)](#exploring-datasets)
- [Data Cleaning](#data-cleaning)
  - [Handling missing data](#handling-missing-data)
  - [Removing duplicates](#removing-duplicates)
- [Saving Cleaned DataFrames](#Saving-Cleaned-DataFrames)

In [2]:
# Import libraries
import os
import pandas as pd

import sys
sys.path.append(os.path.abspath(".."))
import utils as u

In [3]:
# Set display options for better readability in output
#pd.set_option("display.max_colwidth", None)
pd.set_option("display.float_format", '{:,.2f}'.format)

## **Data Overview**

### Loading raw datasets

➤ Load all raw CSV files into individual DataFrames and store them in a dictionary for easier handling and quick access during exploration.

In [4]:
# Load all CSV files from the Brazilian E-Commerce dataset into separate DataFrames
data_path = "../data/raw/"
files = os.listdir(data_path)

customers_raw = pd.read_csv(data_path + "olist_customers_dataset.csv")
geolocation_raw = pd.read_csv(data_path + "olist_geolocation_dataset.csv")
orders_raw = pd.read_csv(data_path + "olist_orders_dataset.csv")
items_raw = pd.read_csv(data_path + "olist_order_items_dataset.csv")
payments_raw = pd.read_csv(data_path + "olist_order_payments_dataset.csv")
reviews_raw = pd.read_csv(data_path + "olist_order_reviews_dataset.csv")
products_raw = pd.read_csv(data_path + "olist_products_dataset.csv")
sellers_raw = pd.read_csv(data_path + "olist_sellers_dataset.csv")
translation_raw = pd.read_csv(data_path + "product_category_name_translation.csv")

# Store all DataFrames in a dictionary for easier looping and inspection
dataframes_raw = {
    "customers": customers_raw,
    "geolocation": geolocation_raw,
    "orders": orders_raw,
    "items": items_raw,
    "payments": payments_raw,
    "reviews": reviews_raw,
    "products": products_raw,
    "sellers": sellers_raw,
    "translation": translation_raw,
}


The following tables are included in the Brazilian E-Commerce dataset:

- `customers`: customer information  
- `geolocation`: geographical coordinates by zip code prefix  
- `orders`: order details including status and timestamps  
- `items`: product-level details for each order  
- `payments`: payment methods, amounts and installment information 
- `reviews`: customer reviews and ratings  
- `products`: product attributes including category and dimensions
- `sellers`: seller information  
- `translation`: Portuguese-to-English product category mapping 

*Note: Original file names such as `olist_customers_dataset.csv` were renamed to simpler identifiers like `customers` for ease of use.*

[🠉 Back to top](#structure-of-the-notebook)

### Exploring Datasets

➤  Summary of all tables using `.shape`, column names, and duplicate counts.

In [5]:
# Call the function to display the summary of all loaded DataFrames
u.summarize_table_shapes(dataframes_raw)

name     rows  columns  \
0    customers    99441        5   
1  geolocation  1000163        5   
2       orders    99441        8   
3        items   112650        7   
4     payments   103886        5   
5      reviews    99224        7   
6     products    32951        9   
7      sellers     3095        4   
8  translation       71        2   

                                        column_names  duplicates  
0  [customer_id, customer_unique_id, customer_zip...           0  
1  [geolocation_zip_code_prefix, geolocation_lat,...      261831  
2  [order_id, customer_id, order_status, order_pu...           0  
3  [order_id, order_item_id, product_id, seller_i...           0  
4  [order_id, payment_sequential, payment_type, p...           0  
5  [review_id, order_id, review_score, review_com...           0  
6  [product_id, product_category_name, product_na...           0  
7  [seller_id, seller_zip_code_prefix, seller_cit...           0  
8  [product_category_name, product_category_name_...           0

➤  Quick sampling of 5 rows from each table for visual inspection

In [6]:
# Display a random sample of 5 rows from each DataFrame for a quick visual inspection
for name, df in dataframes_raw.items():
    print(f'{name.capitalize()}:')
    display(df.sample(5))
    print("-"*130)

Customers:


customer_id                customer_unique_id  \
918    eacb0c4f63256a598da605d10fd5b17b  2e438ed54223fcca39c725b9a3e6376f   
19747  bdc6a9fb78c7bc3357e405a272342bc2  4c7ba37a823a6222050a800bf3465e7b   
66410  9a99e708648d1215092a52e22376c0a3  0d5a795edaf891c13a4f12fd395c244e   
83904  5cba34aee6ef2cdb57fe60ed3a7cdd8b  28a116b50f7b859f0f5b113b87912738   
493    2173c184771a8a19f6fc87d8b3dea9e0  5e65571d0f33dc17c099484ae3f25a4a   

       customer_zip_code_prefix   customer_city customer_state  
918                       78890         sorriso             MT  
19747                      4661       sao paulo             SP  
66410                     20231  rio de janeiro             RJ  
83904                     38610            unai             MG  
493                       88054   florianopolis             SC

----------------------------------------------------------------------------------------------------------------------------------
Geolocation:


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
711263                        51011            -8.10           -34.88   
503085                        27281           -22.49           -44.10   
680495                        42700           -12.88           -38.31   
580977                        33913           -19.78           -44.02   
540477                        29845           -18.54           -40.30   

          geolocation_city geolocation_state  
711263              recife                PE  
503085       volta redonda                RJ  
680495    lauro de freitas                BA  
580977  ribeirão das neves                MG  
540477       boa esperanca                ES

----------------------------------------------------------------------------------------------------------------------------------
Orders:


order_id                       customer_id  \
91106  bc281606ec7787151f45acecb5d48598  46edbd3283961bf0b23db785bae94b55   
94933  8fa4c1bf3c1b27d1349e2aa934aed53f  3109eab8c772e5f8761f25a92045442b   
35975  c0cbad6a3c8fdde69b705b0016769afe  40d27646b244c9b66fd821672dde0d4c   
51370  30b971d10c8ea1517964eb4b7f720f2c  98a99b42d086d32ac68864925c5ab7c4   
92793  09b6daf7c780e789668ee3b7f4bda1c5  032f5eb3f82038a0490fd6870bf131dd   

      order_status order_purchase_timestamp    order_approved_at  \
91106    delivered      2018-06-14 20:12:13  2018-06-14 20:40:41   
94933    delivered      2018-07-04 10:26:00  2018-07-05 16:35:45   
35975    delivered      2018-08-06 12:18:36  2018-08-07 12:15:24   
51370    delivered      2017-09-01 08:14:59  2017-09-01 09:25:10   
92793    delivered      2018-06-29 18:20:57  2018-07-05 16:04:53   

      order_delivered_carrier_date order_delivered_customer_date  \
91106          2018-06-15 14:25:00           2018-06-21 19:02:59   
94933          2018-07-11 12:34:00           2018-07-12 23:32:09   
35975          2018-08-07 13:24:00           2018-08-13 21:36:44   
51370          2017-09-01 14:53:02           2017-09-08 22:06:44   
92793          2018-07-03 13:08:00           2018-07-09 10:48:30   

      order_estimated_delivery_date  
91106           2018-07-04 00:00:00  
94933           2018-07-19 00:00:00  
35975           2018-08-17 00:00:00  
51370           2017-09-21 00:00:00  
92793           2018-07-26 00:00:00

----------------------------------------------------------------------------------------------------------------------------------
Items:


order_id  order_item_id  \
28765  417753e7b9376497030b2f497535bf5c              1   
96701  db747e7d69671f32e8e4ea7fdc1266ed              1   
45222  66c0ba5488a397208076ece07748d0fa              1   
63483  90f6254d7d1be81c8b501dec6668d86a              1   
37414  54e988b8e764a9967caeacfd27f8942b              1   

                             product_id                         seller_id  \
28765  a92930c327948861c015c919a0bcb4a8  6560211a19b47992c3666cc44a7e94c0   
96701  407eeaa1c0dcd776f6c2ad41db4c62e5  0873d9f8f36123f8d910f4760e788cfb   
45222  8f3a1c2358138bedfb98645db449110c  fa1c13f2614d7b5c4749cbc52fecda94   
63483  4a41bfa11528f2c8133c8fceb4494e84  1d29dfba02015238dfbe2449a5eaa361   
37414  f532ec9a55ee4e466e1c8fc913563ad5  3be634553519fb6536a03e1358e9fdc7   

       shipping_limit_date  price  freight_value  
28765  2018-06-14 00:31:59  78.00          15.65  
96701  2018-07-25 12:55:25 966.90          28.70  
45222  2017-09-12 04:10:32 279.90          15.71  
63483  2017-11-19 23:46:47  52.90          12.50  
37414  2017-11-22 22:31:06  39.90          17.60

----------------------------------------------------------------------------------------------------------------------------------
Payments:


order_id  payment_sequential payment_type  \
50924  f315f8fbe6489c99c9e99dcbe52dace4                   1  credit_card   
30067  2a873cd8f2f02ea9bf84a5ae74970e20                   1  credit_card   
44598  d7f1fc795c94efd3a00141a5f8a2c9d7                   1   debit_card   
19890  3ffd96ca0884bf472feae91fe6f1de09                   1  credit_card   
93754  74e98d2b01c0fd8e1357032c441f09a8                   1  credit_card   

       payment_installments  payment_value  
50924                     4         136.64  
30067                     9         187.96  
44598                     1          59.97  
19890                     5         361.45  
93754                    10         259.88

----------------------------------------------------------------------------------------------------------------------------------
Reviews:


review_id                          order_id  \
18314  bccfba8de128c359ea2c63d46f335741  fa80f0d51be5bc219ebc8193038c17b2   
58506  d6d23f21047f3330e3b94e5fc51bf2ac  6ab838d74d628b6f618844ef50200969   
7880   c12fa4ccff7028919957f7ae8d6219b5  69bff117da8229936027f2bb1024866b   
4434   92b03cdcb722b15187e3bee2b33c1590  94e5d86b81534a5f11b5b57af7852aed   
6819   6a0d9e81b659725c740754b9a92da96a  6a6d60aba373f10c15519e023de52de1   

       review_score review_comment_title  \
18314             5                  NaN   
58506             4                  NaN   
7880              5                  NaN   
4434              5                  NaN   
6819              5                  NaN   

                                  review_comment_message review_creation_date  \
18314           Uma pena que o produto nao funcionou !!!  2018-05-15 00:00:00   
58506                                                NaN  2017-12-09 00:00:00   
7880                                                 NaN  2018-07-07 00:00:00   
4434                                                 NaN  2018-02-18 00:00:00   
6819   O produto é excelente, já tinha uma e adquiri ...  2017-09-09 00:00:00   

      review_answer_timestamp  
18314     2018-05-16 00:02:36  
58506     2017-12-13 23:27:49  
7880      2018-07-10 01:15:58  
4434      2018-02-18 18:36:05  
6819      2017-09-10 01:13:56

----------------------------------------------------------------------------------------------------------------------------------
Products:


product_id product_category_name  \
29237  a20f56d082d66a0caeab963592f3cad5          beleza_saude   
6578   c0d133f59b1b9482e5b66b398ea60a69       cama_mesa_banho   
1755   726838a1b5bdf32e2fc8010f0807bd46          market_place   
13909  b1ab3619ae093935fbb0e0033ac2046f      moveis_decoracao   
925    148d27957e1653a2e9c8c3771f5a5a0f    relogios_presentes   

       product_name_lenght  product_description_lenght  product_photos_qty  \
29237                45.00                      549.00                1.00   
6578                 54.00                      235.00                1.00   
1755                 19.00                      767.00                1.00   
13909                56.00                      169.00                1.00   
925                  60.00                      628.00                2.00   

       product_weight_g  product_length_cm  product_height_cm  \
29237            650.00              19.00              12.00   
6578           1,900.00              34.00               8.00   
1755             200.00              22.00              24.00   
13909          7,300.00              40.00              40.00   
925              200.00              16.00              11.00   

       product_width_cm  
29237             13.00  
6578              32.00  
1755              24.00  
13909             40.00  
925               11.00

----------------------------------------------------------------------------------------------------------------------------------
Sellers:


seller_id  seller_zip_code_prefix  \
351   671585f5a2af58b6e276bc01003c0d2b                   80510   
1588  fad44952713764836814be105382aee5                   21340   
2356  1bb2bdb95f4841f1bba2c0d2cd83d3c9                    1257   
1620  d2df77822373f19934e87694d8640767                   86706   
1239  5741dc2bf5044c34d0698681c5081bed                   12226   

              seller_city seller_state  
351              curitiba           PR  
1588       rio de janeiro           RJ  
2356            sao paulo           SP  
1620            arapongas           PR  
1239  sao jose dos campos           SP

----------------------------------------------------------------------------------------------------------------------------------
Translation:


product_category_name product_category_name_english
31          market_place                  market_place
5          esporte_lazer                sports_leisure
14            brinquedos                          toys
9     relogios_presentes                 watches_gifts
47         moveis_quarto             furniture_bedroom

----------------------------------------------------------------------------------------------------------------------------------


➤ Column-wise overview including dtypes, missing values, and unique counts.

In [7]:
# Quick overview of column properties (dtypes, missing values, uniques) for all DataFrames
u.column_overview(dataframes_raw)

Customers:


dtype  non_null  missing_n  missing_%  uniques_n  \
customer_id               object     99441          0       0.00      99441   
customer_unique_id        object     99441          0       0.00      96096   
customer_zip_code_prefix   int64     99441          0       0.00      14994   
customer_city             object     99441          0       0.00       4119   
customer_state            object     99441          0       0.00         27   

                                                                    uniques  
customer_id               [06b8999e2fba1a1fbc88172c00ba8bc7, 18955e83d33...  
customer_unique_id        [861eff4711a542e4b93843c6dd7febb0, 290c77bc529...  
customer_zip_code_prefix  [14409, 9790, 1151, 8775, 13056, 89254, 4534, ...  
customer_city             [franca, sao bernardo do campo, sao paulo, mog...  
customer_state            [SP, SC, MG, PR, RJ, RS, PA, GO, ES, BA, MA, M...

----------------------------------------------------------------------------------------------------------------------------------
Geolocation:


dtype  non_null  missing_n  missing_%  \
geolocation_zip_code_prefix    int64   1000163          0       0.00   
geolocation_lat              float64   1000163          0       0.00   
geolocation_lng              float64   1000163          0       0.00   
geolocation_city              object   1000163          0       0.00   
geolocation_state             object   1000163          0       0.00   

                             uniques_n  \
geolocation_zip_code_prefix      19015   
geolocation_lat                 717360   
geolocation_lng                 717613   
geolocation_city                  8011   
geolocation_state                   27   

                                                                       uniques  
geolocation_zip_code_prefix  [1037, 1046, 1041, 1035, 1012, 1047, 1013, 102...  
geolocation_lat              [-23.54562128115268, -23.54608112703553, -23.5...  
geolocation_lng              [-46.63929204800168, -46.64482029837157, -46.6...  
geolocation_city             [sao paulo, são paulo, sao bernardo do campo, ...  
geolocation_state            [SP, RN, AC, RJ, ES, MG, BA, SE, PE, AL, PB, C...

----------------------------------------------------------------------------------------------------------------------------------
Orders:


dtype  non_null  missing_n  missing_%  \
order_id                       object     99441          0       0.00   
customer_id                    object     99441          0       0.00   
order_status                   object     99441          0       0.00   
order_purchase_timestamp       object     99441          0       0.00   
order_approved_at              object     99281        160       0.16   
order_delivered_carrier_date   object     97658       1783       1.79   
order_delivered_customer_date  object     96476       2965       2.98   
order_estimated_delivery_date  object     99441          0       0.00   

                               uniques_n  \
order_id                           99441   
customer_id                        99441   
order_status                           8   
order_purchase_timestamp           98875   
order_approved_at                  90733   
order_delivered_carrier_date       81018   
order_delivered_customer_date      95664   
order_estimated_delivery_date        459   

                                                                         uniques  
order_id                       [e481f51cbdc54678b7cc49136f2d6af7, 53cdb2fc8bc...  
customer_id                    [9ef432eb6251297304e76186b10a928d, b0830fb4747...  
order_status                   [delivered, invoiced, shipped, processing, una...  
order_purchase_timestamp       [2017-10-02 10:56:33, 2018-07-24 20:41:37, 201...  
order_approved_at              [2017-10-02 11:07:15, 2018-07-26 03:24:27, 201...  
order_delivered_carrier_date   [2017-10-04 19:55:00, 2018-07-26 14:31:00, 201...  
order_delivered_customer_date  [2017-10-10 21:25:13, 2018-08-07 15:27:45, 201...  
order_estimated_delivery_date  [2017-10-18 00:00:00, 2018-08-13 00:00:00, 201...

----------------------------------------------------------------------------------------------------------------------------------
Items:


dtype  non_null  missing_n  missing_%  uniques_n  \
order_id              object    112650          0       0.00      98666   
order_item_id          int64    112650          0       0.00         21   
product_id            object    112650          0       0.00      32951   
seller_id             object    112650          0       0.00       3095   
shipping_limit_date   object    112650          0       0.00      93318   
price                float64    112650          0       0.00       5968   
freight_value        float64    112650          0       0.00       6999   

                                                               uniques  
order_id             [00010242fe8c5a6d1ba2dd792cb16214, 00018f77f2f...  
order_item_id        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...  
product_id           [4244733e06e7ecb4970a6e2683c13e61, e5f2d52b802...  
seller_id            [48436dade18ac8b2bce089ec2a041202, dd7ddc04e1b...  
shipping_limit_date  [2017-09-19 09:45:35, 2017-05-03 11:05:13, 201...  
price                [58.9, 239.9, 199.0, 12.99, 199.9, 21.9, 19.9,...  
freight_value        [13.29, 19.93, 17.87, 12.79, 18.14, 12.69, 11....

----------------------------------------------------------------------------------------------------------------------------------
Payments:


dtype  non_null  missing_n  missing_%  uniques_n  \
order_id               object    103886          0       0.00      99440   
payment_sequential      int64    103886          0       0.00         29   
payment_type           object    103886          0       0.00          5   
payment_installments    int64    103886          0       0.00         24   
payment_value         float64    103886          0       0.00      29077   

                                                                uniques  
order_id              [b81ef226f3fe1789b1e8b2acac839d17, a9810da8291...  
payment_sequential    [1, 2, 4, 5, 3, 8, 6, 7, 10, 11, 17, 19, 27, 1...  
payment_type          [credit_card, boleto, voucher, debit_card, not...  
payment_installments  [8, 1, 2, 3, 6, 5, 4, 10, 7, 12, 9, 13, 15, 24...  
payment_value         [99.33, 24.39, 65.71, 107.78, 128.45, 96.12, 8...

----------------------------------------------------------------------------------------------------------------------------------
Reviews:


dtype  non_null  missing_n  missing_%  uniques_n  \
review_id                object     99224          0       0.00      98410   
order_id                 object     99224          0       0.00      98673   
review_score              int64     99224          0       0.00          5   
review_comment_title     object     11568      87656      88.34       4527   
review_comment_message   object     40977      58247      58.70      36159   
review_creation_date     object     99224          0       0.00        636   
review_answer_timestamp  object     99224          0       0.00      98248   

                                                                   uniques  
review_id                [7bc2406110b926393aa56f80a40eba40, 80e641a11e5...  
order_id                 [73fc7af87114b39712e6da79b0a377eb, a548910a1c6...  
review_score                                               [4, 5, 1, 3, 2]  
review_comment_title     [nan, recomendo, Super recomendo, Não chegou m...  
review_comment_message   [nan, Recebi bem antes do prazo estipulado., P...  
review_creation_date     [2018-01-18 00:00:00, 2018-03-10 00:00:00, 201...  
review_answer_timestamp  [2018-01-18 21:46:59, 2018-03-11 03:05:13, 201...

----------------------------------------------------------------------------------------------------------------------------------
Products:


dtype  non_null  missing_n  missing_%  \
product_id                   object     32951          0       0.00   
product_category_name        object     32341        610       1.85   
product_name_lenght         float64     32341        610       1.85   
product_description_lenght  float64     32341        610       1.85   
product_photos_qty          float64     32341        610       1.85   
product_weight_g            float64     32949          2       0.01   
product_length_cm           float64     32949          2       0.01   
product_height_cm           float64     32949          2       0.01   
product_width_cm            float64     32949          2       0.01   

                            uniques_n  \
product_id                      32951   
product_category_name              73   
product_name_lenght                66   
product_description_lenght       2960   
product_photos_qty                 19   
product_weight_g                 2204   
product_length_cm                  99   
product_height_cm                 102   
product_width_cm                   95   

                                                                      uniques  
product_id                  [1e9e8ef04dbcff4541ed26657ea517e5, 3aa071139cb...  
product_category_name       [perfumaria, artes, esporte_lazer, bebes, util...  
product_name_lenght         [40.0, 44.0, 46.0, 27.0, 37.0, 60.0, 56.0, 57....  
product_description_lenght  [287.0, 276.0, 250.0, 261.0, 402.0, 745.0, 127...  
product_photos_qty          [1.0, 4.0, 2.0, 3.0, 5.0, 9.0, 6.0, nan, 7.0, ...  
product_weight_g            [225.0, 1000.0, 154.0, 371.0, 625.0, 200.0, 18...  
product_length_cm           [16.0, 30.0, 18.0, 26.0, 20.0, 38.0, 70.0, 40....  
product_height_cm           [10.0, 18.0, 9.0, 4.0, 17.0, 5.0, 24.0, 8.0, 1...  
product_width_cm            [14.0, 20.0, 15.0, 26.0, 13.0, 11.0, 44.0, 40....

----------------------------------------------------------------------------------------------------------------------------------
Sellers:


dtype  non_null  missing_n  missing_%  uniques_n  \
seller_id               object      3095          0       0.00       3095   
seller_zip_code_prefix   int64      3095          0       0.00       2246   
seller_city             object      3095          0       0.00        611   
seller_state            object      3095          0       0.00         23   

                                                                  uniques  
seller_id               [3442f8959a84dea7ee197c632cb2df15, d1b65fc7deb...  
seller_zip_code_prefix  [13023, 13844, 20031, 4195, 12914, 20920, 5532...  
seller_city             [campinas, mogi guacu, rio de janeiro, sao pau...  
seller_state            [SP, RJ, PE, PR, GO, SC, BA, DF, RS, MG, RN, M...

----------------------------------------------------------------------------------------------------------------------------------
Translation:


dtype  non_null  missing_n  missing_%  \
product_category_name          object        71          0       0.00   
product_category_name_english  object        71          0       0.00   

                               uniques_n  \
product_category_name                 71   
product_category_name_english         71   

                                                                         uniques  
product_category_name          [beleza_saude, informatica_acessorios, automot...  
product_category_name_english  [health_beauty, computers_accessories, auto, b...

----------------------------------------------------------------------------------------------------------------------------------


➤ Quick statistical overview of all numeric columns in each raw table to spot any unusual values or patterns.

In [8]:
# Summarize basic statistics of all numeric columns for each DataFrame in the dictionary
u.describe_numeric(dataframes_raw)

Customers:


count      mean       std      min       25%  \
customer_zip_code_prefix 99,441.00 35,137.47 29,797.94 1,003.00 11,347.00   

                               50%       75%       max  
customer_zip_code_prefix 24,416.00 58,900.00 99,990.00

----------------------------------------------------------------------------------------------------------------------------------
Geolocation:


count      mean       std      min  \
geolocation_zip_code_prefix 1,000,163.00 36,574.17 30,549.34 1,001.00   
geolocation_lat             1,000,163.00    -21.18      5.72   -36.61   
geolocation_lng             1,000,163.00    -46.39      4.27  -101.47   

                                  25%       50%       75%       max  
geolocation_zip_code_prefix 11,075.00 26,530.00 63,504.00 99,990.00  
geolocation_lat                -23.60    -22.92    -19.98     45.07  
geolocation_lng                -48.57    -46.64    -43.77    121.11

----------------------------------------------------------------------------------------------------------------------------------
Orders:
No numeric columns to describe.
----------------------------------------------------------------------------------------------------------------------------------
Items:


count   mean    std  min   25%   50%    75%      max
order_item_id 112,650.00   1.20   0.71 1.00  1.00  1.00   1.00    21.00
price         112,650.00 120.65 183.63 0.85 39.90 74.99 134.90 6,735.00
freight_value 112,650.00  19.99  15.81 0.00 13.08 16.26  21.15   409.68

----------------------------------------------------------------------------------------------------------------------------------
Payments:


count   mean    std  min   25%    50%    75%  \
payment_sequential   103,886.00   1.09   0.71 1.00  1.00   1.00   1.00   
payment_installments 103,886.00   2.85   2.69 0.00  1.00   1.00   4.00   
payment_value        103,886.00 154.10 217.49 0.00 56.79 100.00 171.84   

                           max  
payment_sequential       29.00  
payment_installments     24.00  
payment_value        13,664.08

----------------------------------------------------------------------------------------------------------------------------------
Reviews:


count  mean  std  min  25%  50%  75%  max
review_score 99,224.00  4.09 1.35 1.00 4.00 5.00 5.00 5.00

----------------------------------------------------------------------------------------------------------------------------------
Products:


count     mean      std  min    25%    50%  \
product_name_lenght        32,341.00    48.48    10.25 5.00  42.00  51.00   
product_description_lenght 32,341.00   771.50   635.12 4.00 339.00 595.00   
product_photos_qty         32,341.00     2.19     1.74 1.00   1.00   1.00   
product_weight_g           32,949.00 2,276.47 4,282.04 0.00 300.00 700.00   
product_length_cm          32,949.00    30.82    16.91 7.00  18.00  25.00   
product_height_cm          32,949.00    16.94    13.64 2.00   8.00  13.00   
product_width_cm           32,949.00    23.20    12.08 6.00  15.00  20.00   

                                75%       max  
product_name_lenght           57.00     76.00  
product_description_lenght   972.00  3,992.00  
product_photos_qty             3.00     20.00  
product_weight_g           1,900.00 40,425.00  
product_length_cm             38.00    105.00  
product_height_cm             21.00    105.00  
product_width_cm              30.00    118.00

----------------------------------------------------------------------------------------------------------------------------------
Sellers:


count      mean       std      min      25%  \
seller_zip_code_prefix 3,095.00 32,291.06 32,713.45 1,001.00 7,093.50   

                             50%       75%       max  
seller_zip_code_prefix 14,940.00 64,552.50 99,730.00

----------------------------------------------------------------------------------------------------------------------------------
Translation:
No numeric columns to describe.
----------------------------------------------------------------------------------------------------------------------------------


[🠉 Back to top](#structure-of-the-notebook)

## **Data Cleaning**

➤ Copy raw DataFrames into a new working dictionary to preserve the original data before cleaning.

In [9]:
# Create a new dictionary with copies of all raw DataFrames
dataframes = u.copy_dataframes(dataframes_raw, exclude=["geolocation"]) # exclude 'geolocation', which is not used in the analysis

[🠉 Back to top](#structure-of-the-notebook)

### Handling missing data

➤ Missing product-related information

In [10]:
# Copy the 'products' DataFrame from the dictionary for further processing
products = dataframes["products"].copy()

# Display the number of missing values in each column of the 'products' DataFrame
products.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [11]:
# Count the number of rows where any important product-related column is missing
products_missing_cols = [
    "product_category_name",
    "product_name_lenght",
    "product_description_lenght",
    "product_photos_qty"
]

number_missing_all = products[products_missing_cols].isna().any(axis=1).sum()

# Print the number and percentage of affected rows
print(
    f"{number_missing_all} rows have missing values in key product-related columns "
    f"({number_missing_all / products.shape[0]:.1%} of the products table)."
)

610 rows have missing values in key product-related columns (1.9% of the products table).


In [12]:
# Drop rows where all key product-related columns are missing (category, name length, description length, and photo count)
products.dropna(subset=products_missing_cols, how="all", inplace=True)

# Display the number of missing values remaining in each column
products.isna().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              1
product_length_cm             1
product_height_cm             1
product_width_cm              1
dtype: int64

Rows missing `product_category_name` (610 in total) are dropped, as this column is critical for analyzing product-level trends in customer satisfaction. Without it, meaningful grouping and interpretation are not possible.

These rows also lack values in other key descriptive columns — such as `product_name_lenght`, `product_description_lenght`, and `product_photos_qty` — which are relevant for understanding how product presentation may affect customer perception. Since these fields are simultaneously missing, and no reliable imputation method is available, we exclude these rows entirely.

Other missing fields, such as product dimensions and weight, are not directly relevant to our current analysis. However, to ensure the dataset remains reusable for future projects, we will impute the missing values for `product_weight_g`, `product_length_cm`, `product_height_cm`, and `product_width_cm` using median values from products in the same category and a similar price range.

In [13]:
# Define the columns that describe the product's physical dimensions
size_cols = [
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm",
]

# View rows where at least one of the size columns has missing values
products[products[size_cols].isna().any(axis=1)]

product_id product_category_name  \
8578  09ff539a621711667c43eba6a3bd8466                 bebes   

      product_name_lenght  product_description_lenght  product_photos_qty  \
8578                60.00                      865.00                3.00   

      product_weight_g  product_length_cm  product_height_cm  product_width_cm  
8578               NaN                NaN                NaN               NaN

There is now only one row in the dataset with missing values (`product_id` equal to `'09ff539a621711667c43eba6a3bd8466'`), and it happens to be missing all four size-related columns. This product belongs to the `'bebe'` category. Despite this being an isolated case, we will write the following code in a generalized way to ensure it can be reused for other projects or datasets with more missing values.

In [14]:
# Merge median item price into the products table to enable price-based imputation
items = dataframes["items"].copy()
price_by_product = items.groupby("product_id")["price"].median().reset_index()
products_price = products.merge(price_by_product, on="product_id", how="left")

In [15]:
# Create price bins (quartiles) within each product category, based on median price
# This helps identify similar products by both category and price level
products_price["price_bin"] = products_price.groupby("product_category_name")[
    "price"
].transform(lambda x: pd.qcut(x, q=4, duplicates="drop"))

In [16]:
# Impute missing size values using the median for each (category, price_bin) group
for col in size_cols:
    # Calculate the group-specific median for the current column
    group_medians = products_price.groupby(["product_category_name", "price_bin"])[
        col
    ].transform("median")
    # Fill missing values in the current column using the group medians
    products_price[col] = products_price[col].fillna(group_medians)

In [17]:
# Print the number of missing values remaining in each column for verification
# Note: 'price_bin' may have missing values if a product belongs to a category with only one product,
# since quartile-based binning (qcut) cannot be applied in such cases.
print(products_price.isna().sum())
# Verify that the one product with missing values was successfully imputed
products_price[products_price["product_id"] == '09ff539a621711667c43eba6a3bd8466']

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
price                         0
price_bin                     1
dtype: int64


product_id product_category_name  \
8412  09ff539a621711667c43eba6a3bd8466                 bebes   

      product_name_lenght  product_description_lenght  product_photos_qty  \
8412                60.00                      865.00                3.00   

      product_weight_g  product_length_cm  product_height_cm  \
8412          8,200.00              42.00              32.00   

      product_width_cm    price        price_bin  
8412             40.00 1,934.00  (159.9, 3899.0]

In [18]:
# Remove the helper columns and save the cleaned/imputed data back to 'products'
products = products_price.drop(columns=["price", "price_bin"])

[🠉 Back to top](#structure-of-the-notebook)

➤ Missing English translation for product category names

Another issue identified during the data overview stage is that the `products` table contains two category names that are missing from the `translation` table. To ensure consistency in our analysis, we will merge the two tables to add English category names and manually fill in the missing translations.

In [19]:
# Merge products with the translation table
translation = dataframes["translation"].copy()

products_translated = pd.merge(products, translation, how="left", on="product_category_name")

# Identify product categories that are missing English translations
categories_missing_translation = products_translated[
    products_translated["product_category_name_english"].isna()
]["product_category_name"].unique()

print(f"Categories missing English translation: {categories_missing_translation}")

Categories missing English translation: ['pc_gamer' 'portateis_cozinha_e_preparadores_de_alimentos']


The category name `'pc_gamer'` can remain unchanged.
`'portateis_cozinha_e_preparadores_de_alimentos'` refers to kitchen appliances and food preparation tools and will be translated as `'kitchen_appliances_and_food_processors'`.

In [20]:
# Manual translation for missing categories
category_translation_dict = {
    "pc_gamer": "pc_gamer",
    "portateis_cozinha_e_preparadores_de_alimentos": "kitchen_appliances_and_food_processors",
}

# Fill in missing translations using the manual dictionary
products_translated["product_category_name_english"] = products_translated[
    "product_category_name_english"
].fillna(products_translated["product_category_name"].map(category_translation_dict))


Now that we have handled the missing values in the `products` table and added English translations for all product categories, we can update the table in the dictionary. The `translation` table is no longer needed and will be removed.  
Columns related to product dimensions, weight, and the original Portuguese category names will be dropped at a later stage.

In [21]:
# Update the 'products' table in the dictionary with the translated version
dataframes["products"] = products_translated

# Remove the 'translation' table from the dictionary since it's no longer needed
del dataframes["translation"]

[🠉 Back to top](#structure-of-the-notebook)

### Removing duplicates

During the data overview stage, we found that neither `review_ids` nor `order_ids` are unique in the `reviews` table. However, for a consistent and reliable analysis of customer satisfaction, it is essential to ensure a one-to-one relationship between reviews and orders — meaning each `order_id` should be associated with exactly one `review_id`, and each `review_id` should correspond to only one `order_id`.
This prevents double-counting reviews and ensures that satisfaction scores are accurately linked to individual orders.

➤ Removing duplicates in `reviews_id` column

In [23]:
# Copy the reviews and items tables
reviews = dataframes["reviews"].copy()
items = dataframes["items"].copy()

# Filter rows where the same review_id appears more than once
duplicate_reviews = reviews[reviews.duplicated("review_id", keep=False)].copy()

# Count how many times each review_id appears
review_id_counts = duplicate_reviews["review_id"].value_counts()

# Add the count back to the duplicate_reviews dataframe
duplicate_reviews["review_id_count"] = duplicate_reviews["review_id"].map(review_id_counts)

# Merge with items table to inspect what products are tied to these duplicated reviews
duplicate_reviews_items = pd.merge(duplicate_reviews, items, how="left", on="order_id")

# Report the number of duplicated review_id entries
print(
    f"Number of duplicated review_id entries: {duplicate_reviews.shape[0]} out of {reviews.shape[0]} "
    f"({duplicate_reviews.shape[0] / reviews.shape[0]:.1%} of the reviews dataset)."
)

# Sort by frequency of review_id, then review_id and order_id for easier inspection
duplicate_reviews_items_sorted = duplicate_reviews_items.sort_values(
    ["review_id_count", "review_id", "order_id"], ascending=[False, True, True]
)

# Display top rows with the most duplicated review_ids
duplicate_reviews_items_sorted.head(10)

Number of duplicated review_id entries: 1603 out of 99224 (1.6% of the reviews dataset).


review_id                          order_id  \
739   08528f70f579f0c830189efc523d2182  03310aa823a66056268a3bab36e827fb   
1420  08528f70f579f0c830189efc523d2182  53c71d3953507c6239ff73917ed358c9   
137   08528f70f579f0c830189efc523d2182  7813842ae95e8c497fc0233232ae815a   
1053  0c76e7a547a531e7bf9f0b99cba071c1  16cc0fe71527d13426bdbe29205b2053   
372   0c76e7a547a531e7bf9f0b99cba071c1  3525e0e57f9d276d522d570bd46cb39c   
626   0c76e7a547a531e7bf9f0b99cba071c1  98c977c116f7779360e9fecffd3860b6   
891   1fb4ddc969e6bea80e38deec00393a6f  3c1098cb17277b62cfc709c7a9b500f5   
885   1fb4ddc969e6bea80e38deec00393a6f  afed4265a8b956d840bc032e54dfccd1   
414   1fb4ddc969e6bea80e38deec00393a6f  d6dde74bdeb424af6b660214881b4845   
415   1fb4ddc969e6bea80e38deec00393a6f  d6dde74bdeb424af6b660214881b4845   

      review_score review_comment_title  \
739              1       Produto errado   
1420             1       Produto errado   
137              1       Produto errado   
1053             5                  NaN   
372              5                  NaN   
626              5                  NaN   
891              5                  NaN   
885              5                  NaN   
414              5                  NaN   
415              5                  NaN   

                                 review_comment_message review_creation_date  \
739   Entrega do produto diferente do solicitado\r\n...  2018-08-03 00:00:00   
1420  Entrega do produto diferente do solicitado\r\n...  2018-08-03 00:00:00   
137   Entrega do produto diferente do solicitado\r\n...  2018-08-03 00:00:00   
1053                                                NaN  2017-08-31 00:00:00   
372                                                 NaN  2017-08-31 00:00:00   
626                                                 NaN  2017-08-31 00:00:00   
891                                                 NaN  2017-08-11 00:00:00   
885                                                 NaN  2017-08-11 00:00:00   
414                                                 NaN  2017-08-11 00:00:00   
415                                                 NaN  2017-08-11 00:00:00   

     review_answer_timestamp  review_id_count  order_item_id  \
739      2018-08-06 00:09:52                3            NaN   
1420     2018-08-06 00:09:52                3           1.00   
137      2018-08-06 00:09:52                3            NaN   
1053     2017-09-05 15:27:17                3           1.00   
372      2017-09-05 15:27:17                3           1.00   
626      2017-09-05 15:27:17                3           1.00   
891      2017-08-12 14:35:35                3           1.00   
885      2017-08-12 14:35:35                3           1.00   
414      2017-08-12 14:35:35                3           1.00   
415      2017-08-12 14:35:35                3           2.00   

                            product_id                         seller_id  \
739                                NaN                               NaN   
1420  17d98fc630d23a628ec1130f8d69d1d2  f9244d45189d3a3605499abddeade7d5   
137                                NaN                               NaN   
1053  5c821d665e58b08092e830f81c12cc37  abc015ebac1d283a323ca9db35ede4b9   
372   fba6f88e10230dd0ba718aec962064ae  d91fb3b7d041e83b64a00a3edfb37e4f   
626   e4aa3f3e3209009166c6b79afa8204bc  1127b7f2594683f2510f1c2c834a486b   
891   d800be06cad2262f48d1c2982fd31189  cfb1a033743668a192316f3c6d1d2671   
885   45a19bf9dddef525a5881b62c01f164c  004c9cd9d87a3c30c522c48c4fc07416   
414   2475f9f6b36857e1a35292cf761ffa48  4a3ca9315b744ce9f8e9374361493884   
415   548e5bfe28edceab6b51fa707cc9556f  da8622b14eb17ae2831f4ac5b9dab84a   

      shipping_limit_date    price  freight_value  
739                   NaN      NaN            NaN  
1420  2018-08-13 20:55:31 3,109.99          74.35  
137                   NaN      NaN            NaN  
1053  2017-09-01 20:44:59    89.00          12.12  
372   2017-09-01 20:44:58    18.00 

The table above shows examples of `review_id`s that appear multiple times, each linked to a different `order_id`. These cases likely involve multi-item purchases — either of the same or different products — that should have been recorded under a single `order_id`, but were incorrectly split while sharing the same review.

Since this duplication affects only 1.6% of the reviews and introduces ambiguity into the mapping between `orders` and `reviews`, we consider it safer to remove these inconsistent entries from the dataset to preserve the integrity of the analysis.

In [24]:
# Identify review_ids that appear more than once
duplicated_ids = reviews["review_id"][reviews["review_id"].duplicated(keep=False)]

# Remove rows with duplicated review_ids
reviews_no_duplicated_review_id = reviews[~reviews["review_id"].isin(duplicated_ids)]

# Print summary
print(
    f"Number of rows in updated reviews table: {reviews_no_duplicated_review_id.shape[0]}"
)
print(
    f"Number of unique review_ids: {reviews_no_duplicated_review_id['review_id'].nunique()}"
)


Number of rows in updated reviews table: 97621
Number of unique review_ids: 97621


[🠉 Back to top](#structure-of-the-notebook)

➤ Removing duplicates in `order_id` column

In [25]:
# Filter rows where the same order_id appears more than once
duplicate_orders = reviews_no_duplicated_review_id[
    reviews_no_duplicated_review_id.duplicated("order_id", keep=False)
].copy()

# Count how many times each order_id appears
order_id_counts = duplicate_orders["order_id"].value_counts()

# Add the count as a new column for easier inspection
duplicate_orders["order_id_count"] = duplicate_orders["order_id"].map(order_id_counts)

# Merge with items table to inspect which products are tied to these duplicated orders
duplicate_orders_items = pd.merge(duplicate_orders, items, how="left", on="order_id")

# Report the number of duplicated order_id entries
print(
    f"Number of duplicated order_id entries: {duplicate_orders.shape[0]} out of {reviews_no_duplicated_review_id.shape[0]} "
    f"({duplicate_orders.shape[0] / reviews_no_duplicated_review_id.shape[0]:.1%} of the reviews dataset after removing review_id duplicates)."
)

# Sort by frequency of order_id, then by order_id and review_id
duplicate_orders_items_sorted = duplicate_orders_items.sort_values(
    ["order_id_count", "order_id", "review_id"], ascending=[False, True, True]
)

# Display top rows with the most duplicated order_ids
duplicate_orders_items_sorted.head(10)

Number of duplicated order_id entries: 182 out of 97621 (0.2% of the reviews dataset after removing review_id duplicates).


review_id                          order_id  \
44   2eab0b2e6f5bded4d9b0b2afcfdf4534  02e0b68852217f5715fb9cc885829454   
88   fa493ead9b093fb0fa6f7d4905b0ef3b  02e0b68852217f5715fb9cc885829454   
160  405eb2ea45e1dbe2662541ae5b47e2aa  03c939fd7fd3b38f8485a0f95798f1f6   
26   b04ed893318da5b863e878cd3d0511df  03c939fd7fd3b38f8485a0f95798f1f6   
103  c793b25686e244588d4fc71964c95a2c  04f1827088d972a62224f5203a071500   
107  f543cfa959c3289703f144a32113bd78  04f1827088d972a62224f5203a071500   
135  7a49d6129a979bbbb500cb14853d2714  056bfadd41b8600ad5ecfef2ac132188   
68   acf3f5b4d93975335ee6a76654cd495b  056bfadd41b8600ad5ecfef2ac132188   
101  1b729a0ff69b1b85e6105a45072cdf9e  0749426d1c48fe5943cbdf1316ace0aa   
93   6d9236aeca36f0a36200fdd41e61e28c  0749426d1c48fe5943cbdf1316ace0aa   

     review_score review_comment_title  \
44              4                  NaN   
88              4                  NaN   
160             3                  NaN   
26              3                  NaN   
103             1                  NaN   
107             5                  NaN   
135             4                  NaN   
68              4                  NaN   
101             1                  NaN   
93              1                  NaN   

                                review_comment_message review_creation_date  \
44                             Gostei chegou rapidinho  2017-12-03 00:00:00   
88                         Gostei e entregou rapidinho  2017-12-01 00:00:00   
160  Seria ótimo se tivesem entregue os 3 (três) pe...  2018-03-06 00:00:00   
26   Um ponto negativo que achei foi a cobrança de ...  2018-03-20 00:00:00   
103                                                NaN  2018-01-07 00:00:00   
107                                                NaN  2017-12-30 00:00:00   
135                                                NaN  2017-12-05 00:00:00   
68                                                 NaN  2017-12-05 00:00:00   
101  Um década pra entregar o produto. Horrível a e...  2018-02-09 00:00:00   
93      Pessima entrega demora um século pra entregar   2018-02-10 00:00:00   

    review_answer_timestamp  order_id_count  order_item_id  \
44      2017-12-03 21:56:37               2              1   
88      2017-12-03 21:57:31               2              1   
160     2018-03-06 19:50:32               2              1   
26      2018-03-21 02:28:23               2              1   
103     2018-01-07 11:14:04               2              1   
107     2018-01-03 10:40:06               2              1   
135     2017-12-09 13:53:55               2              1   
68      2017-12-09 13:53:28               2              1   
101     2018-02-10 12:55:50               2              1   
93      2018-02-10 12:48:27               2              1   

                           product_id                         seller_id  \
44   2028bf1b01cafb2d2b1901fca4083222  cc419e0650a3c5ba77189a1882b7556a   
88   2028bf1b01cafb2d2b1901fca4083222  cc419e0650a3c5ba77189a1882b7556a   
160  ab5da1daa941470d14366f4e76a99dd2  d91fb3b7d041e83b64a00a3edfb37e4f   
26   ab5da1daa941470d14366f4e76a99dd2  d91fb3b7d041e83b64a00a3edfb37e4f   
103  daf9301b3e3756d45250765852ed0dfd  004c9cd9d87a3c30c522c48c4fc07416   
107  daf9301b3e3756d45250765852ed0dfd  004c9cd9d87a3c30c522c48c4fc07416   
135  2bb67f2cd3ff60400e0bad33aee72d2a  daeb5653dd96c1b11860f72209795012   
68   2bb67f2cd3ff60400e0bad33aee72d2a  daeb5653dd96c1b11860f72209795012   
101  9bb8ca338e5588c361e34eae02e8fad6  620c87c171fb2a6dd6e8bb4dec959fc6   
93   9bb8ca338e5588c361e34eae02e8fad6  620c87c171fb2a6dd6e8bb4dec959fc6   

     shipping_limit_date  price  freight_value  
44   2017-12-05 06:31:54  56.99          11.90  
88   2017-12-05 06:31:54  56.99          11.90  
160  2018-03-02 09:35:46  26.00          16.11  
26   2018-03-02 09:35:46  26.00          16.11  
103  2017-12-18 20:50:26 159.69          18.37  
107  2017-12-18 20:50:26 159.69          18.37  
135  2017-11-30 1

In many cases, the same `order_id` is associated with multiple `review_id`s, which might differ in content or score. These appear to be edited versions of the same customer review, mistakenly recorded as separate entries.
To ensure that each order is represented by only one review, we keep only the latest available review per `order_id`, based on the `review_answer_timestamp` field, which offers the most precise timestamp available.

In [26]:
# Convert review_answer_timestamp to datetime
reviews_no_duplicated_review_id.loc[:, "review_answer_timestamp"] = pd.to_datetime(
    reviews_no_duplicated_review_id["review_answer_timestamp"], errors="coerce"
)

In [27]:
# Sort and keep the latest review per order_id
reviews_cleaned = reviews_no_duplicated_review_id.sort_values(
    "review_answer_timestamp"
).drop_duplicates(subset="order_id", keep="last")

In [28]:
# Print total number of rows and number of unique review_ids and order_ids in cleaned table
print(f"Number of rows in cleaned reviews table: {reviews_cleaned.shape[0]}")
print("Number of unique review_ids and order_ids:")
print(reviews_cleaned[['review_id', "order_id"]].nunique())


Number of rows in cleaned reviews table: 97530
Number of unique review_ids and order_ids:
review_id    97530
order_id     97530
dtype: int64


Now that both `review_id` and `order_id` columns are unique, we can update the `dataframes` dictionary with the `reviews_cleaned` table.

In [29]:
# Update the 'reviews' table in the dictionary with the cleaned version
dataframes["reviews"] = reviews_cleaned

[🠉 Back to top](#structure-of-the-notebook)

## **Saving Cleaned DataFrames**

In [30]:
# Save cleaned dataframes as .csv files
save_path = "../data/cleaned"

for name, df in dataframes.items():
    file_path = os.path.join(save_path, f"{name}.csv")
    df.to_csv(file_path, index=False)